src/redux/sagas/login.ts

```diff
@@ -40,19 +41,26 @@ function* loginStatusWatcherSaga() {
 
     if (user) {
       // login procedure
-      yield put({ type: WalletActionType.WALLET_REQUESTED, uid: user.uid });
-      const walletAction: WalletRetrievedAction = yield take(WalletActionType.WALLET_RETRIEVED);
-      const { wallet } = walletAction;
-      const player = yield fetchOrCreatePlayer(wallet.address, user.displayName);
+      yield fork(walletSaga, user.uid);
 
-      // playerHeartbeatThread = yield fork(playerHeartbeatSaga, wallet.address);
-      yield put(MessageAction.subscribeMessages(wallet.address));
-      applicationThread = yield fork(applicationControllerSaga, wallet);
+      const { success, failure } = yield race({
+        success: take(walletActions.INITIALIZATION_SUCCESS),
+        failure: call(delay, 2000),
+      });
+
+      if (failure) { throw new Error('Wallet initialization timed out'); }
+
+      const address = (success as walletActions.InitializationSuccess).address;
+
+      const player = yield fetchOrCreatePlayer(address, user.displayName);
+
+      yield put(MessageAction.subscribeMessages(address));
+
+      applicationThread = yield fork(applicationControllerSaga, address);
```

Comment:
adding some comments here would be helpful (seems like a lot going on at first glance)